web-ui/src/streaming/import/random-data/index.tsx

```diff
@@ -0,0 +1,194 @@
+// Displays a form with random settings to configure what values are generated
+// for a given field in a table.
+
+import React, { useCallback, useEffect, useMemo, useState } from 'react'
+import { Autocomplete, FormHelperText, Grid, TextField } from '@mui/material'
+import { Controller, useWatch } from 'react-hook-form'
+import dayjs, { Dayjs } from 'dayjs'
+
+import { useCustomForm } from '../hooks/useCustomForm'
+import { Field } from 'src/types/manager'
+import { getValueParser } from 'src/types/ddl'
+import { IRngGenMethod, columnTypeToRngOptions, getRngMethodByName } from './generators'
+import { ExampleDisplay } from './ExampleDisplay'
+import { StoredFieldSettings } from '../RngSettingsDialog'
+import { StoreSettingsFn } from '../ImportToolbar'
+
+// The names of potential form field parameters.
+export enum FieldNames {
+  BOOL_CONST = 'bool_const',
+  TRUE_PCT = 'true_pct',
+  VALUE = 'value',
+  MIN = 'min',
+  MAX = 'max',
+  MU = 'mu',
+  SIGMA = 'sigma',
+  LAMBDA = 'lambda',
+  ALPHA = 'alpha',
+  BETA = 'beta',
+  TIME = 'time',
+  TIME2 = 'time2',
+  DATE_RANGE = 'date_range'
+}
+
+export const RngFieldSettings = (props: {
+  field: Field
+  fieldSettings: StoredFieldSettings | undefined
+  setSettings: StoreSettingsFn
+}) => {
+  const { field, fieldSettings, setSettings } = props
+
+  const defaultValues: Partial<Record<FieldNames, any>> = useMemo(() => {
+    return {
+      bool_const: false,
+      true_pct: '0.5',
+      value: '0',
+      min: '0',
+      max: '100',
+      mu: '100',
+      sigma: '10',
+      lambda: '0.1',
+      alpha: '0.1',
+      beta: '0.1',
+      time: dayjs(new Date(2023, 1, 1, 0, 0)),
+      time2: dayjs(new Date(2023, 12, 1, 23, 59, 59)),
+      date_range: [dayjs(new Date(2023, 1, 1, 0, 0)), dayjs(new Date(2023, 12, 1, 23, 59, 59))]
+    }
+  }, [])
+
+  // The current random generation method that was selected.
+  const [selectedMethod, setSelectedMethod] = useState<IRngGenMethod | null>(
+    fieldSettings ? getRngMethodByName(fieldSettings.method, field.columntype) : null
+  )
+  // An random example value that is generated if a method is selected.
+  const [example, setExample] = useState<boolean | number | string | object | Dayjs | undefined>(undefined)
+  const [parsedExample, setParsedExample] = useState<boolean | number | string | undefined>(undefined)
+
+  // Instantiate a react hook form.
+  //
+  // We use a custom hook that allows us to change the validation schema on the
+  // fly.
+  const {
+    control,
+    updateSchema,
+    clearErrors,
+    formState: { errors, isValid, isDirty }
+  } = useCustomForm({
+    mode: 'onChange',
+    schema: selectedMethod?.validationSchema?.(field.columntype),
```

Comment:
is this syntax some react extension?

Reply:
I don't think so

Reply:
tried something similar in a typescript playground and it failed. what does it mean?